src/builder.ts

```diff
@@ -0,0 +1,146 @@
+import { z } from "zod";
+import { grab } from "./grab";
+import type { Selection } from "./grab";
+
+type Query = string;
+type Payload<T> = { schema: T; query: Query };
+
+export class BaseResult<T> {
+  query: string;
+  schema: T;
+
+  constructor({ query, schema }: Payload<T>) {
+    this.query = query;
+    this.schema = schema;
+  }
+
+  public value(): Payload<T> {
+    return { schema: this.schema, query: this.query };
+  }
+}
+
+/**
+ * Unknown, comes out of pipe and is starting point for queries.
+ */
+export class UnknownResult extends BaseResult<z.ZodUnknown> {
```

Comment:
Not 100% sure it's a good idea, but after looking at your classes, it appears you could refactor and define `UnknownResult` by extending `EntityResult`, in the same way `UnknownArrayResult` extends `ArrayResult`.

Reply:
Another great catch! That doesn't seem to choke any tests, and reduces even more code. Dope!